### ISSUE TITLE
Trying to implement CBAM with yolov5s

### ISSUE THREAD

**Issue by Sanjana9922** [ultralytics/yolov5#11785]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello 
I am trying to implement CBAM with yolov5s so, after changing few files in yolov5 like the common.py , yolo.py and yolov5s.yaml file it is still not taking the CBAM layer. I am not getting any error but CBAM is not implementing hence I am stuck there 

**COMMON.PY**
class Conv(nn.Module):
    # Standard convolution with args(ch_in, ch_out, kernel, stride, padding, groups, dilation, activation)
    default_act = nn.SiLU()  # default activation

    def __init__(self, c1, c2, k=1, s=1, p=None, g=1, d=1, act=True):
        super().__init__()
        self.conv = nn.Conv2d(c1, c2, k, s, autopad(k, p, d), groups=g, dilation=d, bias=False)
        self.bn = nn.BatchNorm2d(c2)
        self.act = self.default_act if act is True else act if isinstance(act, nn.Module) else nn.Identity()

    def forward(self, x):
        return self.act(self.bn(self.conv(x)))

class ChannelAttention(nn.Module):
    def __init__(self, in_planes, ratio=16):
        super(ChannelAttention, self).__init__()
        self.avg_pool = nn.AdaptiveAvgPool2d(1)
        self.max_pool = nn.AdaptiveMaxPool2d(1)
        self.fc1 = nn.Conv2d(in_planes, in_planes // ratio, 1, bias=False)
        self.relu = nn.ReLU()
        self.fc2 = nn.Conv2d(in_planes // ratio, in_planes, 1, bias=False)
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        avg_out = self.fc2(self.relu(self.fc1(self.avg_pool(x))))
        max_out = self.fc2(self.relu(self.fc1(self.max_pool(x))))
        out = self.sigmoid(avg_out + max_out)
        return torch.mul(x, out)

## attenstion mechanism block

class SpatialAttention(nn.Module):
    def __init__(self, kernel_size=7):
        super(SpatialAttention, self).__init__()
        assert kernel_size in (3, 7), 'kernel size must be 3 or 7'
        padding = 3 if kernel_size == 7 else 1
        self.conv = nn.Conv2d(2, 1, kernel_size, padding=padding, bias=False)
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        avg_out = torch.mean(x, dim=1, keepdim=True)
        max_out, _ = torch.max(x, dim=1, keepdim=True)
        out = torch.cat([avg_out, max_out], dim=1)
        out = self.sigmoid(self.conv(out))
        return torch.mul(x, out)

class CBAM(nn.Module):
    def __init__(self, c1, c2, n=1, shortcut=True, g=1, e=0.5):
        super(CBAM, self).__init__()
        c_ = int(c2 * e)
        self.cv1 = Conv(c1, c_, 1, 1)
        self.cv2 = Conv(c1, c_, 1, 1)
        self.cv3 = Conv(2 * c_, c2, 1, 1)
        self.m = nn.Sequential(*[Bottleneck(c_, c_, shortcut, g, e=1.0) for _ in range(n)])
        self.channel_attention = ChannelAttention(c2, 16)
        self.spatial_attention = SpatialAttention(7)

    def forward(self, x):
        return self.spatial_attention(self.channel_attention(self.cv3(torch.cat((self.m(self.cv1(x)), self.cv2(x)), dim=1))))

**YOLO.PY**
n = n_ = max(round(n * gd), 1) if n > 1 else n  # depth gain
        if m in {Conv, GhostConv, Bottleneck, GhostBottleneck, SPP, SPPF, DWConv, MixConv2d, Focus, CrossConv,
                 BottleneckCSP, C3, C3TR, C3SPP, C3Ghost, nn.ConvTranspose2d, DWConvTranspose2d, C3x, CBAM}:
            c1, c2 = ch[f], args[0]
            if c2 != no:  # if not output
                c2 = make_divisible(c2 * gw, 8)

            args = [c1, c2, *args[1:]]
            if m in {BottleneckCSP, C3, C3TR, C3Ghost, C3x, CBAM}:
                args.insert(2, n)  # number of repeats
                n = 1
        elif m is nn.BatchNorm2d:

Yolov5s.yaml
# YOLOv5 🚀 by Ultralytics, GPL-3.0 license

# Parameters
nc: 80  # number of classes
depth_multiple: 0.33  # model depth multiple
width_multiple: 0.50  # layer channel multiple
anchors:
  - [10,13, 16,30, 33,23]  # P3/8
  - [30,61, 62,45, 59,119]  # P4/16
  - [116,90, 156,198, 373,326]  # P5/32

# YOLOv5 v6.0 backbone
backbone:
# [from, number, module, args]
  [[-1, 1, Conv, [64, 6, 2, 2]],  # 0-P1/2
   [-1, 1, Conv, [128, 3, 2]],  # 1-P2/4
   [-1, 3, CBAM, [128]],
   [-1, 3, C3, [128]],
   [-1, 1, Conv, [256, 3, 2]],  # 3-P3/8
   [-1, 6, CBAM, [256]],
   [-1, 6, C3, [256]],
   [-1, 1, Conv, [512, 3, 2]],  # 5-P4/16
   [-1, 9, CBAM, [512]],
   [-1, 9, C3, [512]],
   [-1, 1, Conv, [1024, 3, 2]],  # 7-P5/32
   [-1, 3, CBAM, [1024]],         ##added SE 
   [-1, 1, SPPF, [1024, 5]],  # 9
  ]
# YOLOv5 v6.0 head
head:
  [-1, 3, C3, [256, False]],  # 17 (P3/8-small)

   [-1, 1, Conv, [256, 3, 2]],
   [[-1, 15], 1, Concat, [1]],  # cat head P4    
   [[-1, 14], 1, Concat, [1]],  # cat head P4    
   [-1, 3, C3, [512, False]],  # 20 (P4/16-medium)

   [-1, 1, Conv, [512, 3, 2]],
   [[-1, 11], 1, Concat, [1]],  # cat head P5
   [[-1, 10], 1, Concat, [1]],  # cat head P5
   [-1, 3, C3, [1024, False]],  # 23 (P5/32-large)

   [[18, 21, 24], 1, Detect, [nc, anchors]],  # Detect(P3, P4, P5)
   [[17, 20, 23], 1, Detect, [nc, anchors]],  # Detect(P3, P4, P5)
  

I NEED SOME GUIDANCE

**Comment by Sanjana9922** [comment]
![image](https://github.com/ultralytics/yolov5/assets/136984184/1cb541bf-64ac-44b5-9cb1-dfcdcac0b648)

After changing also I am not getting the CBAM layer

**Comment by glenn-jocher** [comment]
@Sanjana9922 the CBAM layer is not currently implemented in YOLOv5. If you would like to add this layer to your YOLOv5 model, you will need to modify the code accordingly. You can refer to the original CBAM paper or other implementations of the CBAM layer to understand how it works and make the necessary changes to the YOLOv5 codebase. Please note that this will require some familiarity with PyTorch and deep learning architectures.

**Comment by Sanjana9922** [comment]
class ChannelAttention(nn.Module):
    def __init__(self, in_planes, ratio=16):
        super(ChannelAttention, self).__init__()
        self.avg_pool = nn.AdaptiveAvgPool2d(1)
        self.max_pool = nn.AdaptiveMaxPool2d(1)
        self.f1 = nn.Conv2d(in_planes, in_planes // ratio, 1, bias=False)
        self.relu = nn.ReLU()
        self.f2 = nn.Conv2d(in_planes // ratio, in_planes, 1, bias=False)
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        avg_out = self.f2(self.relu(self.f1(self.avg_pool(x))))
        max_out = self.f2(self.relu(self.f1(self.max_pool(x))))
        out = self.sigmoid(avg_out + max_out)
        return out


class SpatialAttention(nn.Module):
    def __init__(self, kernel_size=7):
        super(SpatialAttention, self).__init__()
        assert kernel_size in (3, 7), 'kernel size must be 3 or 7'
        padding = 3 if kernel_size == 7 else 1
  
        self.conv = nn.Conv2d(2, 1, kernel_size, padding=padding, bias=False)
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        # 1*h*w
        avg_out = torch.mean(x, dim=1, keepdim=True)
        max_out, _ = torch.max(x, dim=1, keepdim=True)
        x = torch.cat([avg_out, max_out], dim=1)
        #2*h*w
        x = self.conv(x)
        #1*h*w
        return self.sigmoid(x)


class CBAM(nn.Module):
    # CSP Bottleneck with 3 convolutions
    def __init__(self, c1, c2, ratio=16, kernel_size=7):  # ch_in, ch_out, number, shortcut, groups, expansion
        super(CBAM, self).__init__()
        self.channel_attention = ChannelAttention(c1, ratio)
        self.spatial_attention = SpatialAttention(kernel_size)

    def forward(self, x):
        out = self.channel_attention(x) * x
        # c*h*w
        # c*h*w * 1*h*w
        out = self.spatial_attention(out) * out
        return out

Hey @glenn-jocher is this the correct way to update the CBAM layer in COMMON.PY

**Comment by glenn-jocher** [comment]
Hi @Sanjana9922, thank you for your question!

The code you provided seems to be the implementation of the CBAM layer in PyTorch. To update the CBAM layer in `common.py` of YOLOv5, you can follow these steps:

1. Add the necessary import statement for `CBAM` module:
   ```python
   from your_module_path import CBAM
   ```

2. Locate the `Conv` blocks where you want to incorporate the CBAM layer.

3. After the `Conv` block, instantiate the CBAM layer and pass the required parameters. For example, if you want to add the CBAM layer after a `Conv2d` block:
   ```python
   x = Conv2d(in_channels, out_channels, kernel_size, stride, padding)(x)
   x = CBAM(out_channels)(x)
   ```

4. Repeat this process for each `Conv` block where you want to include the CBAM layer.

Remember to replace `from your_module_path import CBAM` with the actual path to your `CBAM` module. Also, make sure to pass the correct input channels and other parameters to the `CBAM` layer based on your specific use case.

I hope this helps! Let me know if you have any further questions or concerns.

**Comment by Sanjana9922** [comment]
Thank you so much for your guidance. I will try and ask any further queries

**Comment by glenn-jocher** [comment]
@Sanjana9922 thank you for your feedback! I'm glad I could assist you. If you have any further questions or need any more help with YOLOv5, please don't hesitate to ask. I'm here to support you.

**Comment by Blessing988** [comment]
> Hi @Sanjana9922, thank you for your question!
> 
> The code you provided seems to be the implementation of the CBAM layer in PyTorch. To update the CBAM layer in `common.py` of YOLOv5, you can follow these steps:
> 
> 1. Add the necessary import statement for `CBAM` module:
>    ```python
>    from your_module_path import CBAM
>    ```
> 2. Locate the `Conv` blocks where you want to incorporate the CBAM layer.
> 3. After the `Conv` block, instantiate the CBAM layer and pass the required parameters. For example, if you want to add the CBAM layer after a `Conv2d` block:
>    ```python
>    x = Conv2d(in_channels, out_channels, kernel_size, stride, padding)(x)
>    x = CBAM(out_channels)(x)
>    ```
> 4. Repeat this process for each `Conv` block where you want to include the CBAM layer.
> 
> Remember to replace `from your_module_path import CBAM` with the actual path to your `CBAM` module. Also, make sure to pass the correct input channels and other parameters to the `CBAM` layer based on your specific use case.
> 
> I hope this helps! Let me know if you have any further questions or concerns.

Is one supposed to modify the yaml file too?

**Comment by glenn-jocher** [comment]
Hi there! You don't necessarily need to modify the YAML file to integrate the CBAM layer into the YOLOv5 model. The YAML file primarily defines the model architecture at a higher level, specifying the sequence and types of layers and modules used. 

However, if you want to explicitly define where the CBAM layers should be included as part of the architecture directly in the YAML, you would then need to adjust the YAML file accordingly. This would involve adding entries for the CBAM layer in the appropriate places within the architecture specification.

If you're just experimenting or adding the CBAM layers programmatically in the Python files (like in `common.py`), then modifying the YAML isn't required. 😊
